### ISSUE TITLE
Rewrite Cython compilation in setup.py

### ISSUE THREAD

**Issue by cdeil** [scikit-hep/probfit#69]
The current `setup.py` isn't written in a good way.

The two main issues I see are:
1. Cython isn't run automatically when Cython files are edited.
2. `python setup.py egg_info` fails if Cython or Numpy aren't installed.

I think the recommended way to fix 1. is to use Cythonize:
http://docs.cython.org/en/latest/src/userguide/source_files_and_compilation.html

For 2. one has to be careful to put the `Cython` and `numpy` imports in if statements or try blocks, so that `python setup.py egg_info` works.

I don't have time to fix this now. Contributions welcome!

Another thing that should be done is to extend the travis-ci test matrix to something similar like
https://github.com/iminuit/iminuit/blob/master/.travis.yml
to e.g. have a build where no Cython and Numpy are installed and `python setup.py egg_info` is run.

**Comment by alexpearce** [comment]
Number 1. has been fixed by #75.

Number 2. is still true, due to the numpy import. We need numpy regardless of whether we have Cython or not, I think, due to the [numpy imports in the Cython files](https://github.com/iminuit/probfit/blob/master/probfit/costfunc.pyx#L3-L4) which get translated to preprocessor imports in the C files:

```C
// Inside costfunc.c, for example
#include "numpy/arrayobject.h"
#include "numpy/ufuncobject.h"
```

The `Extension` objects in the setup.py file need the numpy include path.

I don't know how to get around that then 😕 Other than require users to `pip install numpy` before installing probfit.

**Comment by cdeil** [comment]
The recommended way to write `setup.py` files for Cython projects is to put the Cython (and for us also Numpy) imports in `if-else` or `try-except` blocks.
For iminuit I (or Piti?) put this:
https://github.com/iminuit/iminuit/blob/master/setup.py#L32

There are several efforts to improve Python packaging to use static config files more instead of `setup.py`, but for now, package managers run `python setup.py egg_info` to just get a version number and dependencies and other basic info. So it is really useful to make that work without needing dependencies. `setuptools` is an exception, the Python packaging authority changed the default examples and recommendation to `import setuptools` in `setup.py` instead of `import distutils` a few years ago.

**Comment by alexpearce** [comment]
probfit now does Cython checking in the same way as iminuit, but iminuit doesn't have the numpy problem probfit does (numpy isn't required to build the iminuit C(++) components).

**Comment by alexpearce** [comment]
To be clear: I'm fully in favour if making `egg_info` (and indeed `pip install probfit`) work without having to install numpy first. I'm just not sure how to do it.

**Comment by cdeil** [comment]
Same trick as with Cython: you have to avoid the numpy import and uses in the top-level module in `setup.py`, and instead only conditionally import and use it.

The options are:
- if
- try, except
- a function
- a module in `probfit` that you import from setup.py if it `setup.py` gets too confusing.

I'm not a `setup.py` expert, I would start by looking at `setup.py` of other projects using Cython and Numpy, like pandas, scikit-image, ...
It might be that these "big" projects always have a complex solution.

Maybe also search the Cython mailing list archive and Stackoverflow, and if you don't find anything, ask a "What is the best practice to write a setup.py using Cython / Numpy" question.

Also -- I don't think this is a blocker for the next release. If you don't want to spend more time on it, just assign the issue to me. (but I don't have time for this short-term).

**Comment by cdeil** [comment]
I see the milestone is already "wishlist".
:-)

**Comment by alexpearce** [comment]
I have a feeling that those packages use wheels when you `pip install`. Building _must_ require numpy, whether or not Cython is present. Cython is needed only if one wants to (re)generate the C files.

Is probfit distributed with wheels? Compiled for each platform?

Interestingly, in a new virtualenv I get an import error about numpy if I do `pip install scikit-learn; python -c 'import sklearn'`. But installing pandas in another new virtualenv doesn't give me the same problem, it installs `numpy`.

**Comment by cdeil** [comment]
> Building must require numpy

Yes, `python setup.py build` needs Numpy.
But `python setup.py egg_info` (the command to extract package name, version, depencency and other meta info) should work without Numpy.

I would bet that it does for packages like `scikit-learn` and `pandas`.
Yes, people are starting to produce wheels for many packages.
But you can always find the source tarballs on PyPI, and these are used by many people and projects (e.g. to build conda package, Linux packages, users building from source).

> Is probfit distributed with wheels? 

No.
Honestly, I don't think the userbase is large enough to bother.
And it's technically very savvy people, they have a C compiler and don't mind waiting a minute for `probfit` to compile.

**Comment by alexpearce** [comment]
Ah! I see, OK. I hadn't thought of people using _only_ `egg_info`.

Thank you for being so patient 😅 

So a diff like this should fix that:

```diff
diff --git a/setup.py b/setup.py
index 35f85a1..7bafbb3 100644
--- a/setup.py
+++ b/setup.py
@@ -3,28 +3,33 @@ import os
 from setuptools import setup
 from setuptools.extension import Extension
 
-import numpy as np
-
 try:
-    from Cython.Build import cythonize
-
-    USE_CYTHON = True
+    import numpy as np
 except ImportError:
-    print('Cython is not available; using pre-generated C files')
-    USE_CYTHON = False
+    print('Could not import numpy; C extensions will not be built')
+    USE_NUMPY = False
 
-ext = '.pyx' if USE_CYTHON else '.c'
 extensions = []
-for source_file in glob('probfit/*' + ext):
-    fname, _ = os.path.splitext(os.path.basename(source_file))
-    extensions.append(
-        Extension('probfit.{0}'.format(fname),
-                  sources=['probfit/{0}{1}'.format(fname, ext)],
-                  include_dirs=[np.get_include()])
-    )
+if USE_NUMPY:
+    try:
+        from Cython.Build import cythonize
+
+        USE_CYTHON = True
+    except ImportError:
+        print('Cython is not available; using pre-generated C files')
+        USE_CYTHON = False
+
+    ext = '.pyx' if USE_CYTHON else '.c'
+    for source_file in glob('probfit/*' + ext):
+        fname, _ = os.path.splitext(os.path.basename(source_file))
+        extensions.append(
+            Extension('probfit.{0}'.format(fname),
+                      sources=['probfit/{0}{1}'.format(fname, ext)],
+                      include_dirs=[np.get_include()])
+        )
 
-if USE_CYTHON:
-    extensions = cythonize(extensions)
+    if USE_CYTHON:
+        extensions = cythonize(extensions)
 
 
 def get_version():
```

Summary: try/except the numpy import, and then only try to create `Extension` objects if we have numpy.

**Comment by cdeil** [comment]
In principle, yes.

But I'll need to see the full new `setup.py` and think or locally try through different cases.
Error messages should also be reasonable.

Can you make a PR, so that I can try it out?
And add a build on `travis-ci` where you don't have the optional dependencies, but still run `setup.py`.
I think we have that for `iminuit`, if no, I could point you to other `.travis.yml` where we have such builds in the build matrix.

**Comment by henryiii** [comment]
Thus was fixed in #105.
